pool/pool.go

```diff
@@ -541,25 +543,154 @@ func (p *pool) checkSessionTokenErr(err error, address string) bool {
+	cp, err := p.connection()
+	if err != nil {
+		return err
+	}
+
+	ctx.endpoint = cp.address
+	ctx.client = cp.client
+
+	ctx.key = cfg.key
+	if ctx.key == nil {
+		ctx.key = p.key
+	}
+
+	ctx.sessionDefault = cfg.useDefaultSession
+	ctx.sessionToken = cfg.stoken
+
+	if ctx.sessionToken == nil && ctx.sessionDefault {
+		cacheKey := formCacheKey(ctx.endpoint, ctx.key)
+
+		ctx.sessionToken = p.cache.Get(cacheKey)
+		if ctx.sessionToken == nil {
+			var cliPrm client.CreateSessionPrm
+
+			cliPrm.SetExp(math.MaxUint32)
+
+			cliRes, err := ctx.client.CreateSession(ctx.ctxBase, cliPrm)
+			if err != nil {
+				return fmt.Errorf("default session: %w", err)
+			}
+
+			ctx.sessionToken = sessionTokenForOwner(owner.NewIDFromPublicKey(&ctx.key.PublicKey), cliRes)
+
+			err = ctx.sessionToken.Sign(ctx.key)
+			if err != nil {
+				return fmt.Errorf("sign token: %w", err)
+			}
+
+			_ = p.cache.Put(cacheKey, ctx.sessionToken)
+		}
+	}
+
+	return nil
+}
+
+func (p *pool) PutObject(ctx context.Context, hdr object.Object, payload io.Reader, opts ...CallOption) (*oid.ID, error) {
 	cfg := cfgFromOpts(append(opts, useDefaultSession())...)
-	cp, options, err := p.conn(ctx, cfg)
+
+	var ctxCall callContext
+
+	ctxCall.ctxBase = ctx
+
+	err := p.prepareCallContext(&ctxCall, cfg)
 	if err != nil {
 		return nil, err
 	}
 
-	res, err := cp.client.PutObject(ctx, params, options...)
+	var prm client.PrmObjectPutInit
 
-	if p.checkSessionTokenErr(err, cp.address) && !cfg.isRetry {
-		opts = append(opts, retry())
-		return p.PutObject(ctx, params, opts...)
+	wObj, err := ctxCall.client.ObjectPutInit(ctx, prm)
+	if err != nil {
+		return nil, fmt.Errorf("init writing on API client: %w", err)
 	}
 
+	wObj.UseKey(*ctxCall.key)
+
+	if ctxCall.sessionToken != nil {
+		wObj.WithinSession(*ctxCall.sessionToken)
+	}
+
+	if cfg.btoken != nil {
+		wObj.WithBearerToken(*cfg.btoken)
+	}
+
+	if wObj.WriteHeader(hdr) {
+		sz := hdr.PayloadSize()
+
+		if data := hdr.Payload(); len(data) > 0 {
+			if payload != nil {
+				payload = io.MultiReader(bytes.NewReader(data), payload)
+			} else {
+				payload = bytes.NewReader(data)
+				sz = uint64(len(data))
+			}
+		}
+
+		if payload != nil {
+			const defaultBufferSizePut = 4096 // configure?
+
+			if sz == 0 || sz > defaultBufferSizePut {
+				sz = defaultBufferSizePut
+			}
+
+			buf := make([]byte, sz)
+
+			var n int
+			var ok bool
+
+			for {
+				n, err = payload.Read(buf)
+				if err != nil {
+					if errors.Is(err, io.EOF) {
+						break
+					}
+
+					return nil, fmt.Errorf("read payload: %w", err)
+				}
+
+				ok = wObj.WritePayloadChunk(buf[:n])
+				if !ok {
+					break
+				}
+			}
+		}
+	}
+
+	res, err := wObj.Close()
 	if err != nil { // here err already carries both status and client errors
-		return nil, err
+		p.checkSessionTokenErr(err, ctxCall.endpoint)
+		return nil, fmt.Errorf("client failure: %w", err)
 	}
 
-	return res.ID(), nil
+	id := hdr.ID()
+	if id == nil {
```

Comment:
I don't think we can rely on `hdr.ID()` value ever. We don't know if the object will be extended with required headers on the server size or not. Return ID from response in 100% cases.